snorkel/analysis/utils.py

```diff
@@ -0,0 +1,251 @@
+from typing import List, Optional, Tuple, Union
+
+import numpy as np
+import scipy.sparse as sparse
+import torch
+
+from snorkel.types import ArrayLike
+
+
+def probs_to_preds(probs: np.ndarray) -> np.ndarray:
+    """Convert an array of probabilistic labels into an array of predictions
+
+    Parameters
+    ----------
+    prob
+        A [num_datapoints, num_classes] array of probabilistic labels such that each
+        row sums to 1.
+
+    Returns
+    -------
+    np.ndarray
+        A [num_datapoints, 1] array of predictions (integers in [1, ..., num_classes])
+    """
+    return np.argmax(probs, axis=1) + 1
+
+
+def preds_to_probs(preds: np.ndarray, num_classes: int) -> np.ndarray:
+    """Convert an array of probabilistic labels into an array of predictions
+
+    Parameters
+    ----------
+    pred
+        A [num_datapoints] or [num_datapoints, 1] array of predictions
+
+    Returns
+    -------
+    np.ndarray
+        A [num_datapoints, num_classes] array of probabilistic labels with probability
+        of 1.0 in the column corresponding to the prediction
+    """
+    return np.eye(num_classes)[preds.squeeze() - 1]
+
+
+def arraylike_to_numpy(
+    array_like: ArrayLike, flatten: bool = True, cast_to_int: bool = True
+) -> np.ndarray:
+    """Convert an ArrayLike (e.g., list, tensor, etc.) to a numpy array
+
+    Also optionally flatten [n, 1] arrays to [n] and cast all values to ints.
+    This method is typically used to sanitize labels before use with analysis tools or
+    metrics that expect 1D numpy arrays as inputs.
+
+    Parameters
+    ----------
+    array_like
+        An Arraylike to convert
+    flatten
+        If True, flatten numpy array into a 1D array
+    cast_to_int
+        If True, cast all values to ints
+
+    Returns
+    -------
+    np.ndarray
+        The input converted to an np.ndarray
+
+    Raises
+    ------
+    ValueError
+        Provided input could not be converted to an np.ndarray
+    """
+
+    orig_type = type(array_like)
+
+    # Convert to np.ndarray
+    if isinstance(array_like, np.ndarray):
+        pass
+    elif isinstance(array_like, list):
+        array_like = np.array(array_like)
+    elif isinstance(array_like, sparse.spmatrix):
+        array_like = array_like.toarray()
+    elif isinstance(array_like, torch.Tensor):
+        array_like = array_like.numpy()
+    elif not isinstance(array_like, np.ndarray):
+        array_like = np.array(array_like)
+    else:
+        msg = f"Input of type {orig_type} could not be converted to an np.ndarray"
+        raise ValueError(msg)
+
+    # Correct shape
```

Comment:
In a previous PR, @henryre proposed replacing these with a single `squeeze()` command. The issue is that there are places where we use this to convert golds/preds/probs to numpy before applying metrics to them, and we don't want to flatten and cast the probs.